# Hypothesis testing problems

In [2]:
import numpy as np
import pandas as pd
import scipy.stats as stats
import random

## Exercise 1

You are a nutritionist researching two different types of diets to see if there is a significant difference in weight loss after one month. You choose two random groups of people; one group follows the first diet, and the other follows the second. At the end of the month, you record the weight loss (in kg) for each person:

| Diet 1 | Diet 2 |
|:-------|:-------|
| 2.0 | 3.0 |
| 2.5 | 3.2 |
| 3.0 | 3.1 |
| 2.8 | 2.9 |
| 2.3 | 2.8 |
| 2.7 | 3.0 |
| 2.5 | 3.2 |

With these data, it seeks to answer the following question: Is there a significant difference in average weight loss between people who followed the first diet and those who followed the second diet?

To draw conclusions, follow the points below:

- State the hypothesis: null and alternative hypothesis.
- Perform the test to test the hypothesis. You can use a Student's t-test.
- Analyze the conclusions.

In [3]:
# TODO
diet1 = np.array([2.0, 2.5, 3.0, 2.8, 2.3, 2.7, 2.5])
diet2 = np.array([3.0, 3.2, 3.1, 2.9, 2.8, 3.0, 3.2])

t_statistic, p_value = stats.ttest_ind(diet1, diet2)

print(f"T-statistic: {t_statistic}")
print(f"P-value: {p_value}")

T-statistic: -3.5383407969933938
P-value: 0.004083270191713912


- T test of (-3.538) suggests the weight loss average of Diet1 is lower than the one for Diet2.
- P value of 0.00408 is less than the recommended level of significance of 0.05. With this we can reject the null hypothesis

Conclusion: The recommended diet for optimal weight loss in this exercise is Diet2

## ANOVA

**ANOVA** (*Analysis of Variance*) is a statistical technique used to compare the measures of two or more groups. The idea behind ANOVA is to decompose the total variability in the data into two components: between-group variability and within-group variability:

- **Between-group variability**: This variability refers to the differences between the group means. If this variability is considerably larger than the within-group variability, it could be an indication that at least one of the group means is different.
- **Within-group variability**: This variability refers to the dispersion of the data within each group. If all groups have similar variability, then any noticeable difference in group means could be considered significant.

Hypotheses in ANOVA typically include:

- **Null hypothesis** ($H₀$): The means of all groups are equal.
- **Alternative hypothesis** ($H₁$): At least one of the group means is different.

If the ANOVA test result is significant (e.g., a p-value less than a threshold such as 0.05), this suggests that at least one group mean is different.

In [4]:
diet1 = [2.0, 2.5, 3.0, 2.8, 2.3, 2.7, 2.5]
diet2 = [3.0, 3.2, 3.1, 2.9, 2.8, 3.0, 3.2]

f_statistic, p_value = stats.f_oneway(diet1, diet2)

print(f"F-statistic: {f_statistic}")
print(f"P-value: {p_value}")

F-statistic: 12.519855595667874
P-value: 0.004083270191713883


## Using Pandas

In [5]:
diet1 = [2.0, 2.5, 3.0, 2.8, 2.3, 2.7, 2.5]
diet2 = [3.0, 3.2, 3.1, 2.9, 2.8, 3.0, 3.2]

data = pd.DataFrame({
    'Diet 1': diet1,
    'Diet 2': diet2
})

data_melted = pd.melt(data.reset_index(), id_vars=['index'], value_vars=['Diet 1', 'Diet 2'])
data_melted.columns = ['index', 'Diet', 'Weight Loss']

print(data_melted)

    index    Diet  Weight Loss
0       0  Diet 1          2.0
1       1  Diet 1          2.5
2       2  Diet 1          3.0
3       3  Diet 1          2.8
4       4  Diet 1          2.3
5       5  Diet 1          2.7
6       6  Diet 1          2.5
7       0  Diet 2          3.0
8       1  Diet 2          3.2
9       2  Diet 2          3.1
10      3  Diet 2          2.9
11      4  Diet 2          2.8
12      5  Diet 2          3.0
13      6  Diet 2          3.2


In [6]:
f_statistic, p_value = stats.f_oneway(diet1, diet2)

print(f"F-statistic: {f_statistic}")
print(f"P-value: {p_value}")

F-statistic: 12.519855595667874
P-value: 0.004083270191713883


## Exercise 2

A farmer decides to test three different types of fertilizers to determine if one is superior in terms of corn production. The farmer plants corn on 15 identical plots and uses all three fertilizers (5 plots for each type). At the end of the season, he measures the corn yield (in kg) of each plot, with the following result:

| Fertilizer 1 | Fertilizer 2 | Fertilizer 3 |
|:-------------|:-------------|:-------------|
| 20 | 22 | 24 |
| 21 | 21 | 23 |
| 20 | 23 | 22 |
| 19 | 22 | 23 |
| 20 | 21 | 24 |

With this data, he seeks to answer the following question: Is there a significant difference in average corn yield between the three types of fertilizers?

To help you, follow the points below:

- State the hypothesis: null and alternative hypothesis.
- Perform the ANOVA test.
- Analyze the conclusions.
- If one fertilizer is better than another, how can we know it?

## Hypothesis

Null Hypothesis (H0): The corn yield are equal for all three fertilizers ​

Alternative Hypothesis (H1): At least one of the fertilizers has a different corn yield

In [7]:
# TODO
fertilizer1 = [20, 21, 20, 19, 20]
fertilizer2 = [22, 21, 23, 22, 21]
fertilizer3 = [24, 23, 22, 23, 24]

f_statistic, p_value = stats.f_oneway(fertilizer1, fertilizer2, fertilizer3)

print(f"F-statistic: {f_statistic}")
print(f"P-value: {p_value}")

F-statistic: 20.315789473684188
P-value: 0.00014047824793190475


Conclusion: The F statistics is high. This indicates the mean between the three fertilizers is larger than the variability. Thus one of the fertilizers performs better than the others and with a P value of 0.00014, wich is much lower than the significance level of 0.05 we can reject the Null Hypothesis (H0).

In [8]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd

fertilizer1 = [20, 21, 20, 19, 20]
fertilizer2 = [22, 21, 23, 22, 21]
fertilizer3 = [24, 23, 22, 23, 24]

data = np.array(fertilizer1 + fertilizer2 + fertilizer3)
labels = ['fertilizer1'] * len(fertilizer1) + ['fertilizer2'] * len(fertilizer2) + ['fertilizer3'] * len(fertilizer3)

tukey_result = pairwise_tukeyhsd(data, labels)
tukey_result.summary()

<class 'statsmodels.iolib.table.SimpleTable'>